### DISCUSSION TITLE
We don’t need Hibernate 6 to release Grails 7

### DISCUSSION THREAD

**Discussion by gsartori** [apache/grails-core#13521] [Ideas]
Hibernate 6 improves performances so it’s worth to consider moving ahead and supporting it. At the moment though we should consider the priority on security issues.

Databases are consolidated technologies, having the latest ORM is something that could wait.

Having the latest web framework cannot.

**Comment by jamesfredley** [comment]
@gsartori Do you know of a work around for the following?

Spring Boot 3.x switched to the jakarta.persistence package namespace of JakartaEE 9+, which is supported in Hibernate 6.  

Spring Boot 2.x uses the javax.persistence package namespace of JakartaEE 8 and earlier (JPA 2.2 and earlier) used in Hibernate 5.X.

**Reply by gsartori** [reply]
I have no idea, if it’s just a package name change (meaning the API is the same or compatible) can’t we just fork Hibernate 5 renaming the packages? Better would be creating a “-JAKARTA” version so to not have two separate codebase.

**Comment by ppazos** [comment]
Is Hibernate 6 a blocker to work on the next release of Grails?

**Reply by gsartori** [reply]
I don’t think is blocking, it’s just a lot of work and for my understanding it will not be compatible with the applications already written in Grails 6 since the API has changed and GORM relies on that part of the API. 

I am no expert though so let’s see if anyone more involved in GORM can give us their opinion.

**Comment by gsartori** [comment]
Does anybody know if updating GORM 8 to Groovy 4 and Java 17/21 would be possible staying with Hibernate 5 without adopting Hibernate 6?

**Comment by rainboyan** [comment]
Hibernate started to support JPA 3.0 in version 5.5. All Hibernate artifacts that support the new JPA 3.0 APIs have the suffix “-jakarta” in their name.

```
<dependency>
    <groupId>org.hibernate</groupId>
    <artifactId>hibernate-core-jakarta</artifactId>
    <version>5.5.0.Final</version>
</dependency>
```

https://in.relation.to/2021/06/04/hibernate-is-jakarta-jpa-2/

https://in.relation.to/2021/06/02/hibernate-orm-550-final-release/

**Reply by gsartori** [reply]
So that means we can go with it right? Am I the only one that is not in a hurry of implementing a new GORM based on a new Hibernate API? We are not 300 people, we need to prioritize and decide carefully where to put our effort on.

**Comment by basejump** [comment]
5.6.15 I think should be the initial version of hibernate to chase. Its the latest and we are using it fine with GORM 7 and 8 ( v8 as far as I know) 
as @rainboyan mentioned add the -jakarta suffix for JPA 3. 
see https://central.sonatype.com/search?q=g%3Aorg.hibernate%20v%3A5.6.15.Final&sort=name

<img width="944" alt="image" src="https://github.com/grails/grails-core/assets/187726/726d7e92-8595-4bc7-8020-30595b92a712">

**Comment by gavinking** [comment]
> Hibernate 6 improves performances so it’s worth to consider moving ahead and supporting it. 

I think this _really_ undersells the deep and fundamental advantages that 6/7 offer over 5. Hibernate 6 involved a huge amount of careful rethinking and re-engineering. There's almost no subsystem of the whole sprawling codebase which isn't now more robust and more powerful. The reimplementation of HQL is *alone* enough reason to upgrade.

**Reply by gsartori** [reply]
I guess we all agree it's good to have Hibernate 6, the problem here is "when". We need a glass to be able drink wine. Here we don't have the glass. We need to build one first, than we can pour the wine.

**Reply by gavinking** [reply]
You should also consider that Hibernate 6 is not very backward compatible with Hibernate 5. So by far the best time do such an upgrade is in a major release.

**Reply by gsartori** [reply]
Yep, if we can have Spring Boot 3 + Hibernate 5 for Grails 7 the next major release implementing Hibernate 6 would be Grails 8 :)

**Reply by ppazos** [reply]
I don't understand why the Grails major version depends on the Hibernate major version. If the GORM API for the developer doesn't change, that is, a new GORM includes the new version of Hibernate but doesn't break the API, then it doesn't need a new major version of Grails to accommodate a new major version of Hibernate. A major release should depend mostly on adding breaking changes for exposed APIs. When I say API it's the exposed GORM interface for developers to use. If a feature is added, that is not a breaking change, so it doesn't require a major version release. Though I understand if you have a Grails app in version x.y and it can't work on version x.y+1, I would consider that a breaking change too.

**Reply by gavinking** [reply]
@ppazos I have never used GORM, but that doesn't sound right to me. I very much doubt that GORM completely abstracts Hibernate.

**Reply by ppazos** [reply]
@gavinking In Grails I don't think any persistence feature goes directly to Hibernate without passing through some GORM component. How did you use Grails without using GORM?

**Reply by gavinking** [reply]
> Yep, if we can have Spring Boot 3 + Hibernate 5 for Grails 7 the next major release implementing Hibernate 6 would be Grails 8 :)

Okay, I mean, I have no particular investment in this, but I hope you fully understand that Hibernate 5.6 is already in [limited maintenance mode](https://hibernate.org/community/maintenance-policy/), and the implications of that:

- The last point release of 5.6 was in February 2023, and it's perfectly likely that that was the very last release of 5.6. 
- There will never be a Hibernate 5.7.

**Reply by gavinking** [reply]
> In Grails I don't think any persistence feature goes directly to Hibernate without passing through some GORM component.

I mean, even just a quick scan of the GORM documentation shows me that GORM lets you:

- execute native HQL
- configure things like the second-level cache using Hibernate-defined properties and implementation classes
- use (deprecated) Hibernate-defined id generators

etc.

All this sort of stuff is subject to potential disruption when upgrading H5 -> H6.

> How did you use Grails without using GORM?

I have never used Grails.

**Reply by gsartori** [reply]
> I hope you fully understand that Hibernate 5.6 is already in [limited maintenance mode]

Yes, and we need to give priorities. Upgrading a consolidated DB API is not a priority. Relational databases are mature technologies that won't be impacted by an API that we used for years and reached it's maintenance release. Security, on the other hand, is a priority. 

When I say the word "priority" you have to think at a CTO deciding what technologies to invest on. And a performance optimization is a good thing, but it is never a priority. It can wait a couple of Grails releases.

**Reply by gavinking** [reply]
Haha, OK, carry on, as I said, I have no investment in this at all. Just trying to be helpful.

**Reply by ppazos** [reply]
@gavinking did the HQL language change?

As far as the GORM API doesn't change, I don't see it as a breaking change for Grails, though it could be for the apps created in Grails.

Same for the other two things you mention. I think cache config is GORM DSL.

For the id generators I'm not sure, the only extra generator I used is UUID, and IIRC that is also GORM DSL, not directly hibernate, though there could be other generators. 

In general I still think: if the API doesn't change for the developer, then it's not a breaking change, then it doesn't require a major version release if we follow semver. Note this was more a question than a proposal (just to avoid useless discussions).

**Comment by jamesfredley** [comment]
The latest conversations have been for development to begin around org.hibernate:hibernate-core-jakarta:5.6.15.Final and see how well it will play with Spring Framework 6.1.11 and Spring Boot 3.3.1 within GORM for Grails 7.   Whatever functions and is the simplest path is likely what will ship in with Grails 7.

**Comment by sdelamo** [comment]
[I created a task](https://github.com/grails/grails-core/issues/13550) and added to the [project board](https://github.com/orgs/grails/projects/3).

**Comment by codeconsole** [comment]
I think Hibernate 6 should be skipped completely and and the next Hibernate upgrade target should be [Hibernate 7.0](https://hibernate.org/orm/releases/7.0/). I imagine by the time the upgrade is done, 7 will be released.  The update can be done independently of grails-core.

**Reply by codeconsole** [reply]
Hibernate 7 will be Jakarta EE 11 which won't be supported until end of 2025.  We should just go to [6.6 instead.](https://hibernate.org/orm/releases/6.6/)

We can release Hibernate 7 support with [Grails 8](https://github.com/grails/grails-core/issues/13757)

**Comment by jamesfredley** [comment]
grails-data-hibernate6 work has started: https://github.com/grails/grails-data-hibernate6/branches - see development branches

grails-data-hibernate5 is still seeing material changes:  https://github.com/grails/grails-data-hibernate5/commits/9.0.x/

**hibernate-core-jakarta:5.6.15.Final support in Spring Boot and Spring Framework:**

Spring Boot 3.5.x, like 3.4.x, uses Jakarta persistence 3.1.  3.5.x will be the May 2025 release which should become Grails 7.0.x and Spring will provide OSS support until May 2026.   Spring Boot 3.5.x spring-boot-dependencies points to hibernate 6.6.x: https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-dependencies/3.5.0-M2

https://hibernate.org/orm/releases/5.6/
https://mvnrepository.com/artifact/org.hibernate/hibernate-core-jakarta/5.6.15.Final is using Jakarta persistence 3.0, technically, so we are already cheating the Jakarta Persistence version, but have not seen compatibility issues, to date.

Spring Boot 4.0.x with Spring Framework 7.0.x jumps to Jakarta persistence 3.2, which is the version for Jakarta EE 11:  https://jakarta.ee/specifications/persistence/.  This will be the Nov 2025 Spring Boot release and will likely be the foundation for Grails 8.0.x.
Will hibernate-core-jakarta-5.6.15-final work well enough with Jakarta persistence 3.2, I don't know, but I'd say it's a decent chance.  Hibernate 7.0.x will be the first version of Hibernate to support Jakarta persistence 3.2.  https://in.relation.to/2025/02/13/orm-7beta4/ and https://hibernate.org/orm/releases/7.0/

Given this and how far grails-data-hibernate6 has to go, grails-data-hibernate5 will be around for a while and then there will be a gradual transition to hibernate6 before we stop hibernate5 development all together.  hibernate5 will work with new Grails releases for a minimum of 15 months (May 2026) and if it works with Spring Boot 4.0.x with Spring Framework 7.0.x, then it will work with new Grails releases for some period of years since Spring sticks with an EE release (or two) across major versions.  https://github.com/spring-projects/spring-framework/wiki/Spring-Framework-Versions#javajakarta-ee-versions

Jakarta EE 11 was released June/July 2024 and Spring Framework 7.0.0 will be the Jakarta EE 11 release in Nov 2025.  Jakarta EE 12 is targeting June/July 2026, so Spring Framework 8.0.0 (Jakarta EE 12), I guess, would be targeting Nov 2027.

**Reply by jamesfredley** [reply]
Grails hibernate6 support will likely start with hibernate 6.6 and support Grails 7.0.x initially.  Grails 8, which will be based on Spring 4.0.x, will support Hibernate 7 with Jakarta persistence 3.2, so as early as Nov 2025, earlier if looking at milestones, grails-data-hibernate7 work could begin.   Grails 8 would need to be tested against grails-data-hibernate5, grails-data-hibernate6 (if it's ready) and then considered for grails-data-hibernate7 (later released date).

**Comment by jamesfredley** [comment]
Closing this discussion given we are releasing 7.0.0-RC1 in a few days.  

https://github.com/apache/grails-core/issues/14955 - Hibernate 6.6.x work
https://github.com/apache/grails-core/issues/14956 - Hibernate 7.x.x for Grails 8 issue
